In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.express as px
import plotly.colors as colors
import datetime as dt
from statistics import mode

%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [ ]:
df = pd.read_csv('/content/spam.csv', encoding='latin-1')

In [ ]:
df.sample(5)

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

In [ ]:
df.drop(['Unnamed: 2','Unnamed: 3', 'Unnamed: 4'], axis = 1, inplace = True)

In [ ]:
df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [ ]:
df.isna().any()

v1    False
v2    False
dtype: bool

In [ ]:
df.isna().sum()

v1    0
v2    0
dtype: int64

In [ ]:
df['v2'].nunique()

5169

In [ ]:
df.shape

(5572, 2)

In [ ]:
df['v2'].drop_duplicates(inplace = True)

In [ ]:
df.shape

(5572, 2)

In [ ]:
class_counts = df['v1'].value_counts()

fig = px.bar(class_counts,
             x=class_counts.index,
             y=class_counts.values,
             title='Class Distribution of Spam/Ham',
             labels={'x': 'Spam/Ham', 'y': 'Number of Mails'})

fig.update_layout(xaxis_title='Spam/Ham', yaxis_title='Number of Mails')

fig.show()

In [ ]:
from collections import Counter
import re

import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
all_text = ' '.join(df['v2'].values)
all_text = re.sub(r'http\S+', '', all_text)
all_text = re.sub(r'@\S+', '', all_text)
all_text = re.sub(r'#\S+', '', all_text)

In [ ]:
words = all_text.split()

In [ ]:
stop_words = set(stopwords.words('english'))
words = [word for word in words if not word in stop_words]

In [ ]:
word_counts = Counter(words)
top_words = word_counts.most_common(100)
top_words

[('I', 1466),
 ('u', 692),
 ('2', 453),
 ('call', 376),
 ('get', 326),
 ('U', 299),
 ('ur', 293),
 ("I'm", 286),
 ('&lt;', 276),
 ('You', 263),
 ('4', 249),
 ('.', 235),
 ('go', 234),
 ('know', 224),
 ('like', 222),
 ('got', 204),
 ('come', 198),
 ('?', 187),
 ('...', 163),
 ('want', 157),
 ('Call', 155),
 ('time', 154),
 ('send', 150),
 ('going', 142),
 ('need', 141),
 ('n', 137),
 ("I'll", 137),
 ('How', 137),
 ('still', 134),
 ('If', 133),
 ('one', 132),
 ('But', 131),
 ('No', 126),
 ('text', 126),
 ('Just', 119),
 ('We', 119),
 ('So', 118),
 ('love', 114),
 ('good', 114),
 ('think', 113),
 ('Do', 113),
 ('see', 113),
 ('r', 113),
 ('back', 111),
 ('home', 107),
 ('&', 107),
 ('Ì_', 105),
 ('tell', 104),
 ('Your', 104),
 ('take', 101),
 ('What', 101),
 ('day', 101),
 ('free', 99),
 ('My', 99),
 ('And', 98),
 ('Ok', 97),
 ('me.', 97),
 ('dont', 97),
 ('The', 95),
 ('mobile', 94),
 ('A', 92),
 ("i'm", 91),
 ('FREE', 90),
 ('make', 90),
 ('new', 89),
 ('-', 88),
 ('phone', 88),
 ('late

In [ ]:
top_words = word_counts.most_common(10)
x_values = [word[0] for word in top_words]
y_values = [word[1] for word in top_words]

fig = go.Figure(data=[go.Bar(x=x_values, y=y_values)])
fig.update_layout(title='Most Commonly Used Words', xaxis_title='Word', yaxis_title='Frequency')

fig.show()

# Natural Language Processing
---

In [ ]:
def clean_text(text):
  text = re.sub('<.*?>', '', text)
  text = re.sub('[^a-zA-Z]', ' ', text).lower()
  words = nltk.word_tokenize(text)
  words = [w for w in words if w not in stopwords.words('english')]
  stemmer = PorterStemmer()
  words = [stemmer.stem(w) for w in words]
  text = ' '.join(words)
  return text

In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer

from tqdm.auto import tqdm
import time

In [ ]:
%%time

tqdm.pandas()

df['cleaned_text'] = df['v2'].progress_apply(clean_text)

  0%|          | 0/5572 [00:00<?, ?it/s]

CPU times: user 12.7 s, sys: 1.54 s, total: 14.2 s
Wall time: 15 s


In [ ]:
cv = CountVectorizer(max_features=5000)
X = cv.fit_transform(df['cleaned_text']).toarray()
y = df['v1']

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
lr_classifier = LogisticRegression(random_state=42)
lr_classifier.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [ ]:
y_pred_lr = lr_classifier.predict(X_test)
accuracy_lr = accuracy_score(y_test, y_pred_lr)
print("Logistic Regression Accuracy:", accuracy_lr)

Logistic Regression Accuracy: 0.9775784753363229


In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
dt_classifier = DecisionTreeClassifier(random_state=42)
dt_classifier.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [ ]:
y_pred_dt = dt_classifier.predict(X_test)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
print("Decision Tree Accuracy:", accuracy_dt)

Decision Tree Accuracy: 0.9713004484304932


In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf_classifier = RandomForestClassifier(random_state=42)
rf_classifier.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [ ]:
y_pred_rf = rf_classifier.predict(X_test)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print("Random Forest Accuracy:", accuracy_rf)

Random Forest Accuracy: 0.9766816143497757


In [ ]:
from sklearn.naive_bayes import MultinomialNB

In [ ]:
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train, y_train)

MultinomialNB()

In [ ]:
y_pred = nb_classifier.predict(X_test)
accuracy_nb = accuracy_score(y_test, y_pred)
print("Naive Bayes Accuracy:", accuracy)

Naive Bayes Accuracy: 0.9775784753363229


In [ ]:
from sklearn.svm import SVC

In [ ]:
svm_classifier = SVC(random_state=42)
svm_classifier.fit(X_train, y_train)

SVC(random_state=42)

In [ ]:
y_pred_svm = svm_classifier.predict(X_test)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
print("SVM Accuracy:", accuracy_svm)

SVM Accuracy: 0.9775784753363229


In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
knn_classifier = KNeighborsClassifier()
knn_classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [ ]:
y_pred_knn = knn_classifier.predict(X_test)
accuracy_knn = accuracy_score(y_test, y_pred_knn)
print("KNN Accuracy:", accuracy_knn)

KNN Accuracy: 0.9210762331838565


In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

In [ ]:
gb_classifier = GradientBoostingClassifier(random_state=42)
gb_classifier.fit(X_train, y_train)

GradientBoostingClassifier(random_state=42)

In [ ]:
y_pred_gb = gb_classifier.predict(X_test)
accuracy_gb = accuracy_score(y_test, y_pred_gb)
print("Gradient Boosting Accuracy:", accuracy_gb)

Gradient Boosting Accuracy: 0.967713004484305


In [ ]:
import xgboost as xgb

In [ ]:
y_train_binary = y_train.apply(lambda label: 1 if label == 'spam' else 0)
y_test_binary = y_test.apply(lambda label: 1 if label == 'spam' else 0)

In [ ]:
xgb_classifier = xgb.XGBClassifier(random_state=42)
xgb_classifier.fit(X_train, y_train_binary)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=42, ...)

In [ ]:
y_pred_xgb = xgb_classifier.predict(X_test)
accuracy_xgb = accuracy_score(y_test_binary, y_pred_xgb)
print("XGBoost Accuracy:", accuracy_xgb)

XGBoost Accuracy: 0.9775784753363229


In [ ]:
print("Accuracy Comparison of Machine Learning Algorithms")
print("-----------------------------------------------")
print("{:<25} {:<10}".format('Algorithm', 'Accuracy'))
print("-----------------------------------------------")

algorithm_names = ['Naive Bayes', 'Random Forest', 'SVM', 'Logistic Regression', 'KNN', 'Decision Tree', 'Gradient Boosting', 'XGBoost']
accuracies = [accuracy_nb, accuracy_rf, accuracy_svm, accuracy_lr, accuracy_knn, accuracy_dt, accuracy_gb, accuracy_xgb]

for name, accuracy in zip(algorithm_names, accuracies):
    print("{:<25} {:<10.2f}%".format(name, accuracy * 100))

Accuracy Comparison of Machine Learning Algorithms
-----------------------------------------------
Algorithm                 Accuracy  
-----------------------------------------------
Naive Bayes               97.76     %
Random Forest             97.67     %
SVM                       97.76     %
Logistic Regression       97.76     %
KNN                       92.11     %
Decision Tree             97.13     %
Gradient Boosting         96.77     %
XGBoost                   97.76     %


In [ ]:
algorithm_names = ['Naive Bayes', 'Random Forest', 'SVM', 'Logistic Regression', 'KNN', 'Decision Tree', 'Gradient Boosting', 'XGBoost']
accuracies = [accuracy_nb, accuracy_rf, accuracy_svm, accuracy_lr, accuracy_knn, accuracy_dt, accuracy_gb, accuracy_xgb]

fig = go.Figure()

fig.add_trace(go.Scatter(x=algorithm_names, y=accuracies, mode='lines+markers'))

fig.update_layout(
    title='Accuracy Comparison of Machine Learning Algorithms',
    xaxis_title='Algorithm',
    yaxis_title='Accuracy',
    xaxis_tickangle=-45,
    margin=dict(l=0, r=0, b=0, t=40),
)

fig.show()

The analysis begins with data preprocessing and exploratory data visualization, where you've loaded a dataset containing spam and ham (non-spam) messages. You've removed unnecessary columns, visualized the class distribution using a bar plot, and identified the most commonly used words in the messages.

Following this, you've conducted text preprocessing by cleaning and tokenizing the text data. You've employed various machine learning algorithms to classify messages as spam or ham. The algorithms used include Logistic Regression, Decision Tree, Random Forest, Naive Bayes, Support Vector Machine (SVM), K-Nearest Neighbors (KNN), Gradient Boosting, and XGBoost.

The accuracy comparison of these algorithms reveals the following results:

*   **Naive Bayes:** 97.76%
*   **Random Forest:** 97.67%
*   **SVM:** 97.76%
*   **Logistic Regression**: 97.76%
*   **KNN:** 92.11%
*   **Decision Tree:** 97.13%
*   **Gradient Boosting**: 96.77%
*   **XGBoost:** 97.76%

Based on the analysis, it can be concluded that the Naive Bayes, SVM, Logistic Regression, and XGBoost classifiers are performing consistently well with high accuracy (above 97%). The Random Forest and Decision Tree models are also performing well, although slightly below the top-performing classifiers. The KNN and Gradient Boosting models have slightly lower accuracy compared to the others.

In this particular task of classifying spam and ham messages, the Naive Bayes, SVM, Logistic Regression, and XGBoost classifiers are recommended for their high accuracy. However, it's important to note that accuracy alone might not be the only consideration when selecting a model. Other factors such as model complexity, interpretability, training time, and false positive/negative rates should also be taken into account when making a final decision about which model to use.